### ISSUE TITLE
App Armor denials on Smoke test `Hello World (journald)` on Noble.

### ISSUE THREAD

**Issue by locnnil** [canonical/docker-snap#181]
The smoke-test: 
https://github.com/canonical/docker-snap/blob/a41145de895afc7602cf994ebee855e1c63e7be8/.github/workflows/smoke-test.yml#L112-L119

Fails when running on Noble (but not on Focal and Jammy), as shown here:

Docker Snap **v24.0.5**:
https://github.com/canonical/docker-snap/actions/runs/10738846000

```log
[  210.260771] eth0: renamed from vethfac58ea
[  210.277728] docker0: port 1(vethf07816c) entered blocking state
[  210.277734] docker0: port 1(vethf07816c) entered forwarding state
[  210.333867] vethfac58ea: renamed from eth0
[  210.344540] docker0: port 1(vethf07816c) entered disabled state
[  210.35[222](https://github.com/canonical/docker-snap/actions/runs/10738846000/job/29783725594#step:10:223)0] docker0: port 1(vethf07816c) entered disabled state
[  210.352923] vethf07816c (unregistering): left allmulticast mode
[  210.352928] vethf07816c (unregistering): left promiscuous mode
[  210.352932] docker0: port 1(vethf07816c) entered disabled state
Sep 06 13:10:55 fv-az1693-731 kernel: audit: type=1400 audit(1725628255.588:263): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5269 comm="ps" requested_mask="read" denied_mask="read" peer="/usr/sbin/haveged"
Sep 06 13:10:55 fv-az1693-731 kernel: audit: type=1400 audit(1725628255.590:264): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5269 comm="ps" requested_mask="read" denied_mask="read" peer="rsyslogd"
Sep 06 13:10:55 fv-az1693-731 kernel: audit: type=1400 audit(1725628255.591:265): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5269 comm="ps" requested_mask="read" denied_mask="read" peer="/usr/sbin/chronyd"
Sep 06 13:10:55 fv-az1693-731 kernel: audit: type=1400 audit(1725628255.591:266): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5269 comm="ps" requested_mask="read" denied_mask="read" peer="/usr/sbin/chronyd"
Sep 06 13:10:58 fv-az1693-731 kernel: audit: type=1400 audit(1725628258.948:268): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5590 comm="ps" requested_mask="read" denied_mask="read" peer="/usr/sbin/haveged"
Sep 06 13:10:58 fv-az1693-731 kernel: audit: type=1400 audit(1725628258.950:269): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5590 comm="ps" requested_mask="read" denied_mask="read" peer="rsyslogd"
Sep 06 13:10:58 fv-az1693-731 kernel: audit: type=1400 audit(1725628258.950:270): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5590 comm="ps" requested_mask="read" denied_mask="read" peer="/usr/sbin/chronyd"
Sep 06 13:10:58 fv-az1693-731 kernel: audit: type=1400 audit(1725628258.951:271): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5590 comm="ps" requested_mask="read" denied_mask="read" peer="/usr/sbin/chronyd"
Sep 06 13:11:01 fv-az1693-731 kernel: audit: type=1400 audit(1725628261.245:273): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5713 comm="ps" requested_mask="read" denied_mask="read" peer="/usr/sbin/haveged"
Sep 06 13:11:01 fv-az1693-731 kernel: audit: type=1400 audit(1725628261.247:274): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5713 comm="ps" requested_mask="read" denied_mask="read" peer="rsyslogd"
Sep 06 13:11:01 fv-az1693-731 kernel: audit: type=1400 audit(1725628261.248:275): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5713 comm="ps" requested_mask="read" denied_mask="read" peer="/usr/sbin/chronyd"
Sep 06 13:11:01 fv-az1693-731 kernel: audit: type=1400 audit(1725628261.248:276): apparmor="DENIED" operation="ptrace" class="ptrace" profile="snap.docker.dockerd" pid=5713 comm="ps" requested_mask="read" denied_mask="read" peer="/usr/sbin/chronyd"
Error: Process completed with exit code 1.
```

Docker Snap **v27.2.0** :
https://github.com/canonical/docker-snap/actions/runs/10742009176

**Comment by farshidtz** [comment]
I noticed that docker:log-observe isn't connected here:
https://github.com/canonical/docker-snap/blob/4025064ab31cafde439661c8eb5d3112d30a8eab/.github/workflows/smoke-test.yml#L75-L81

It might be related to this error.

**Comment by locnnil** [comment]
I've tried on [this run](https://github.com/canonical/docker-snap/actions/runs/12641723637), but the error persists.

From my experiments, this is related to the major upgrade of AppArmor from version `3` to version `4` that happened on the transition from Jammy to Noble.

**Comment by farshidtz** [comment]
To reproduce on Noble:
```console
$ docker run --name=test-journald --log-driver=journald bash -c 'echo "hi"'
hi
$ docker logs test-journald
```

**Comment by locnnil** [comment]
Hi @jocado, 

**TL;DR: If you use a docker snap which uses the core24 base, the problem should go away (which is either `core24/beta` channel or `latest/edge`)**

This is a problem with the systemd journald version shipped into Jammy (UC22 base). In short, it's a library incompatibility, Jammy uses an older libsystemd and when you're on Noble you have a newer journal.

```console
root@u2404:~# nsenter -m/run/snapd/ns/docker.mnt                                                     
# journalctl
Journal file /var/log/journal/d930c382a4f54f0788f9ca64139e6d7c/system@e6075808b1154630aed2dec0fc762810-0000000000000397-00062b2c6f094c7d.journal uses an unsupported feature, ignoring file.   
Use SYSTEMD_LOG_LEVEL=debug journalctl --file=/var/log/journal/d930c382a4f54f0788f9ca64139e6d7c/system@e6075808b1154630aed2dec0fc762810-0000000000000397-00062b2c6f094c7d.journal to see the details.
-- No entries --                                 
#       
```

Checking the logs with `SYSTEMD_LOG_LEVEL=debug`:

```console
# SYSTEMD_LOG_LEVEL=debug journalctl --file=/var/log/journal/d930c382a4f54f0788f9ca64139e6d7c/system@e6075808b1154630aed2dec0fc762810-0000000000000397-00062b2c6f094c7d.journal
Journal effective settings seal=no keyed_hash=yes compress=no compress_threshold_bytes=8B
Journal file /var/log/journal/d930c382a4f54f0788f9ca64139e6d7c/system@e6075808b1154630aed2dec0fc762810-0000000000000397-00062b2c6f094c7d.journal has unknown incompatible flags 0x10
Failed to open journal file /var/log/journal/d930c382a4f54f0788f9ca64139e6d7c/system@e6075808b1154630aed2dec0fc762810-0000000000000397-00062b2c6f094c7d.journal: Protocol not supported
mmap cache statistics: 0 context cache hit, 0 window list hit, 1 miss
Failed to open files: Protocol not supported
```

`0x10` it's the problematic flag, introduced here:

https://github.com/systemd/systemd/commit/87413812c92b7a435e94c9c37ed634328d182d4d

So it was introduced in `v252`, while Jammy (22.04) had `v249`.

So, if you switch to the docker channel which uses core24 as base (which is either `core24/beta` channel or `latest/edge`), it should solve the problem.

Let us know if this really fixes the problem, if not it must be something else on the way.

**Comment by farshidtz** [comment]
I can confirm that this can't be reproduced with the snap on latest/beta (core24 based):
```console
$ sudo snap list docker 
Name    Version  Rev   Tracking     Publisher   Notes
docker  28.3.3   3342  latest/beta  canonical✓  -
$ sudo docker run --name=test-journald --log-driver=journald bash -c 'echo "hi"'
hi
$ sudo docker logs test-journald
hi
```

**Comment by locnnil** [comment]
I've tested and by using `latest/edge` (revision **3377**) solves the problem:

```console
$ docker run --name=test-journald --log-driver=journald bash -c 'echo "hi"'
hi
$ docker logs test-journald
hi
```

Same thing for `core24/beta` channel:

```console
$ docker run --name=test-journald --log-driver=journald bash -c 'echo "hi"'
hi
$ docker logs test-journald
hi
```
